### PR TITLE
Libint integration

### PR DESCRIPTION
**Pull request by JensWehner**
Using libint in votca

### CODE REVIEW

**Comment by rubengerritsen** [comment]
@votca-bot format

**Comment by junghans** [comment]
I was trying to backport libint2 to Fedora 32, but it fails due a bug in the gcc Fedora 32 comes with (see https://github.com/evaleev/libint/issues/183). @susilehtola, any idea what we can do here beside waiting for Fedora 33, that has libint2-2.6.0?

**Comment by susilehtola** [comment]
@junghans file a bug against GCC in Fedora 32 with the upstream tracker; they should patch GCC.

**Comment by junghans** [comment]
> @junghans file a bug against GCC in Fedora 32 with the upstream tracker; they should patch GCC.

Good point, thanks! Done: https://bugzilla.redhat.com/show_bug.cgi?id=1881048

**Comment by junghans** [comment]
openSUSE done: https://build.opensuse.org/request/show/834462, now we just have to wait for the next tumbleweed release.

**Comment by junghans** [comment]
And I added libint-2.6.0 to Gentoo in https://github.com/gentoo/gentoo/commit/e61ac24a1e6f7f7c22518017662e26e4f6651e70#diff-b15c1925b645ff28ceadfa30959262dc.

**Comment by junghans** [comment]
> > @junghans file a bug against GCC in Fedora 32 with the upstream tracker; they should patch GCC.
> 
> Good point, thanks! Done: https://bugzilla.redhat.com/show_bug.cgi?id=1881048

@jakubjelinek, is there a chance to get that fixed in Fedora 32?

**Comment by JensWehner** [comment]
> Huh? Didn't we make that target here: https://github.com/votca/votca/blob/master/CMakeModules/BuildLibint.cmake#L15

@junghans Sorry I do not understand

**Comment by junghans** [comment]
> > Huh? Didn't we make that target here: https://github.com/votca/votca/blob/master/CMakeModules/BuildLibint.cmake#L15
> 
> @junghans Sorry I do not understand

I would make the decision for libint  based on the target name not the option

File: CMakeLists.txt [hunk 1]
```diff
@@ -43,9 +43,14 @@ endif()
...
 ########################################################################
 
+
 find_package(PkgConfig REQUIRED)
-set_package_properties(PkgConfig PROPERTIES TYPE RECOMMENDED PURPOSE "Used to detect libxc package")
+set_package_properties(PkgConfig PROPERTIES TYPE RECOMMENDED PURPOSE "Used to detect libxc and libint package")
 pkg_check_modules(LIBXC REQUIRED IMPORTED_TARGET libxc)
+if(NOT TARGET PkgConfig::LIBINT)
+  pkg_check_modules(LIBINT REQUIRED IMPORTED_TARGET libint2>=2.6)
+endif()
```
Comment (JensWehner) [inline CMakeLists.txt:52]: @junghans is that not done here?

Comment (junghans) [reply CMakeLists.txt:52]: This was a comment regarding 220090a2dab526738c5cdfbcb8f4bf2e1ab387ed, which changed that. Not sure if you reverted it afterwards.

Comment (JensWehner) [reply CMakeLists.txt:52]: ahh there was some merge problems going on. I saw your version and I liked it.

**Comment by JensWehner** [comment]
@baumeier this is basically done, do you want to try it out on a system and see if it is slower?
@junghans we have to wait for fedora 33 now, so that the CI does not fail anymore?

**Comment by JensWehner** [comment]
@votca-bot changelog: replaced own integrals with the libint library

**Comment by junghans** [comment]
> @junghans we have to wait for fedora 33 now, so that the CI does not fail anymore?

All builds but `fedora` should work now and `fedora:rawhide` should work as well. Is that the case?

**Comment by JensWehner** [comment]
Expectation management.

**Comment by rubengerritsen** [comment]
> > @junghans we have to wait for fedora 33 now, so that the CI does not fail anymore?
> 
> All builds but `fedora` should work now and `fedora:rawhide` should work as well. Is that the case?

No but we know partly why, fedora:rawhide builds fail  with error: `'LIBINT2_MAX_AM_default1' was not declared in this scope; did you mean 'LIBINT2_MAX_AM_default'?`  and similar error messages

The cause is known see [evaleev/libint#192](https://github.com/evaleev/libint/issues/192), the current answer is that it will be fixed in 2.7.0, the fix is here [evaleev/libint#149](https://github.com/evaleev/libint/pull/149).

Maybe we should make a workaround for that.

**Comment by JensWehner** [comment]
Some distribution packages need "--enable-eri3=1" opensuse and gentoo I think.

**Comment by junghans** [comment]
enable-eri3=1 in openSUSE: https://build.opensuse.org/request/show/841595

**Comment by junghans** [comment]
@mbanck are you still involved with Debian? Could you maybe help us to get libecpint (see #567) included in Debian?

**Comment by mbanck** [comment]
> @mbanck are you still involved with Debian? Could you maybe help us to get libecpint (see #567) included in Debian?

Yes I am, I'll take a look after dinner.

It seems that neither OpenSUSE nor Fedora run the libint tests, I wonder the 32bit breakage on Debian is compiler/toolchain induced.

**Comment by JensWehner** [comment]
@junghans I am confused, I thought the config from fedora has been used on the other systems as well. 

We need as a minimum the configuration as in the https://github.com/votca/votca/blob/master/CMakeModules/BuildLibint.cmake

Otherwise the @rubengerritsen your header fix seems to work. Nice!

**Comment by junghans** [comment]
> @junghans I am confused, I thought the config from fedora has been used on the other systems as well.
> 
> So this requires adding `--enable-eri2=1` as well. Sorry..
> 
> We need as a minimum the configuration as in the https://github.com/votca/votca/blob/master/CMakeModules/BuildLibint.cmake
> 
> Otherwise the @rubengerritsen your header fix seems to work. Nice!

I was wrong the fedora package is already compiled with:
```
        --enable-eri=2 --enable-eri3=2 --enable-eri2=2 \
        --with-eri-max-am=7,5,4 --with-eri-opt-am=3 \
        --with-eri3-max-am=7 --with-eri2-max-am=7 \
        --with-g12-max-am=5 --with-g12-opt-am=3 \
        --with-g12dkh-max-am=5 --with-g12dkh-opt-am=3
```

**Comment by JensWehner** [comment]
Would probably be a good idea to use the same config in gentoo and opensuse.

**Comment by junghans** [comment]
> Would probably be a good idea to use the same config in gentoo and opensuse.

Gentoo: https://github.com/gentoo/gentoo/commit/1e5c68d5369eae6884ba8bbd91548aae56bc2ffb

**Comment by junghans** [comment]
> It seems that neither OpenSUSE nor Fedora run the libint tests, I wonder the 32bit breakage on Debian is compiler/toolchain induced.

@mbanck it seems gentoo has the same problem on x86: https://bugs.gentoo.org/747490#c2

**Comment by mbanck** [comment]
1

> > It seems that neither OpenSUSE nor Fedora run the libint tests, I wonder the 32bit breakage on Debian is compiler/toolchain induced.
> 
> @mbanck it seems gentoo has the same problem on x86: https://bugs.gentoo.org/747490#c2

Are there central build logs from gentoo somewhere to look at? Maybe not, due to the source-only nature?

**Comment by mbanck** [comment]
> > @mbanck are you still involved with Debian? Could you maybe help us to get libecpint (see #567) included in Debian?
> 
> I've uploaded it now, but nowadays the manual processing of new packages can take a while in Debian:  https://salsa.debian.org/debichem-team/libecpint/

@junghans it's in unstable now: https://tracker.debian.org/pkg/libecpint

File: .github/workflows/continuous-integration-workflow.yml [hunk 2]
```diff
@@ -7,7 +7,7 @@ on:
 
 jobs:
   CI:
-    continue-on-error: ${{ matrix.distro == 'ubuntu:devel' || matrix.distro == 'fedora:rawhide' || matrix.continue-on-error == true }}
+    continue-on-error: true
```
Comment (junghans) [inline .github/workflows/continuous-integration-workflow.yml:10]: This line needs to reverted.

Comment (JensWehner) [reply .github/workflows/continuous-integration-workflow.yml:10]: yeah

**Comment by junghans** [comment]
> > > @mbanck are you still involved with Debian? Could you maybe help us to get libecpint (see #567) included in Debian?
> > 
> > 
> > I've uploaded it now, but nowadays the manual processing of new packages can take a while in Debian:  https://salsa.debian.org/debichem-team/libecpint/
> 
> @junghans it's in unstable now: https://tracker.debian.org/pkg/libecpint

@mbanck: Thanks! Looks good, any reason you didn't build it as shared library?

**Comment by junghans** [comment]
> > > > @mbanck are you still involved with Debian? Could you maybe help us to get libecpint (see #567) included in Debian?
> > > 
> > > 
> > > I've uploaded it now, but nowadays the manual processing of new packages can take a while in Debian:  https://salsa.debian.org/debichem-team/libecpint/
> > 
> > 
> > @junghans it's in unstable now: https://tracker.debian.org/pkg/libecpint
> 
> @mbanck: Thanks! Looks good, any reason you didn't build it as shared library?

Ok it was built with `-fPIC` so we should be ok for now.

**Comment by mbanck** [comment]
> @mbanck: Thanks! Looks good, any reason you didn't build it as shared library?

Mostly I did not want to spend too much time on figuring out their API stability engineering; I am always a bit skeptical about CMake projects in this regard, even more so if a static build is the default as seems to be the case here.

But Debian can offer a shared library version later on of course if needed.

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@votca-bot format

**Comment by junghans** [comment]
@votca-bot format

**Comment by JensWehner** [comment]
@baumeier @rubengerritsen please review then we can close the chapter libint.

**Comment by baumeier** [comment]
What does one have to do to make it run in parallel?

Old code (28 threads):
```
... ... 2020-10-31 15:46:26 Calculating Mmn_beta (3-center-repulsion x orbitals)
... ... 2020-10-31 15:51:56 Removed 0 functions from Aux Coulomb matrix to avoid near linear dependencies
... ... 2020-10-31 15:51:56 Calculated Mmn_beta (3-center-repulsion x orbitals)```

Libint (running only on a single thread):
```
... ... 2020-10-31 16:30:1 Calculating Mmn_beta (3-center-repulsion x orbitals)
... ... 2020-10-31 16:40:23 Removed 0 functions from Aux Coulomb matrix to avoid near linear dependencies
... ... 2020-10-31 16:40:23 Calculated Mmn_beta (3-center-repulsion x orbitals)
```

It is promising that it is that fast on a single thread though...

**Comment by junghans** [comment]
@JensWehner codacy complains about something: 
```
src/libxtp/aobasis.cc:232
Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).
table.read(dataVec);
```
same for `src/libxtp/ecpaobasis.cc:163`.

**Comment by junghans** [comment]
In the end we only saved 2k lines of code.......

**Comment by JensWehner** [comment]
@baumeier what do you mean? It should run on 28 cores as well, does the task manager show less threads?

@junghans the codacy warning can be ignored. Concerning the 2k lines of code, I also expected it to be more on the other hand 2k lines is still a bit and we gained a bit of functionality as well.

**Comment by baumeier** [comment]
@JensWehner yes, single thread only

**Comment by JensWehner** [comment]
@baumeier  I guess you ran it with the "-t" option?

Can you post the log?

**Comment by baumeier** [comment]
[dftgwbse_old.log](https://github.com/votca/xtp/files/5469587/dftgwbse_old.log)
[dftgwbse_libint.log](https://github.com/votca/xtp/files/5469588/dftgwbse_libint.log)

**Comment by JensWehner** [comment]
Hmm.. how do you know it is singe threaded?

**Comment by JensWehner** [comment]
I am confused because 80-90% of the calculation should be the gemm calls. Is this also Using cuda?

**Comment by baumeier** [comment]
@JensWehner I take it partially back. The program executes in parallel in the `Calculating Mmn_beta (3-center-repulsion x orbitals)` at first, but then languishes for a long time on a single thread.

**Comment by JensWehner** [comment]
can you try with the new commit?

**Comment by baumeier** [comment]
Took the liberty to time the individual parts of the threecenter `Fill` call:
```
.. ... 2020-10-31 19:49:30 Calculating Mmn_beta (3-center-repulsion x orbitals)  
Starting Fill3cMO Sat Oct 31 19:49:30 2020
done with Fill3cMO Sat Oct 31 19:52:47 2020
done with inv_sqrt Sat Oct 31 19:58:26 2020
done with removedfucntions Sat Oct 31 19:58:26 2020
done with multiply right Sat Oct 31 19:59:25 2020
```
So it's 3 minutes filling the 3C, and 5+ in `inv_sqrt` - checking the current implementation for comparison.

File: src/libxtp/aobasis.cc [hunk 3]
```diff
@@ -65,10 +102,158 @@ void AOBasis::Fill(const BasisSet& bs, const QMMolecule& atoms) {
...
+
+#pragma omp parallel for schedule(dynamic)
+  for (Index s1 = 0; s1 < Index(shells.size()); ++s1) {
+    Index thread_id = OPENMP::getThreadId();
+
+    libint2::Engine& engine = engines[thread_id];
+    const libint2::Engine::target_ptr_vec& buf = engine.results();
+    Index n1 = shells[s1].size();
+
+    for (Index s2 = 0; s2 <= s1; ++s2) {
```
Comment (JensWehner) [inline src/libxtp/aobasis.cc:156]: I think this is the performance bottleneck, we process the small parts first and the large parts at the end. @rubengerritsen

Comment (JensWehner) [reply src/libxtp/aobasis.cc:156]: we should let s2=s1; s2<=shells.size

Comment (JensWehner) [reply src/libxtp/aobasis.cc:156]: but that messes with the ERIs

**Comment by baumeier** [comment]
Now much better:
```
... ... 2020-10-31 22:9:28 Calculating Mmn_beta (3-center-repulsion x orbitals)  
Starting Fill3cMO Sat Oct 31 22:09:28 2020
done with Fill3cMO Sat Oct 31 22:11:39 2020
done with overlap fill Sat Oct 31 22:11:39 2020
done with auxcoulomb fill Sat Oct 31 22:11:39 2020
done with inv_sqrt Sat Oct 31 22:11:52 2020
done with removedfucntions Sat Oct 31 22:11:52 2020
done with multiply right Sat Oct 31 22:12:53 2020
```

**Comment by JensWehner** [comment]
to explain what happened. libint2.hpp includes <eigen/core> so if we include libint2.hpp before our xtp/eigen.h, eigen stuff is compiled without mkl support in that file. So basically libint2.hpp should always be included last.

**Comment by JensWehner** [comment]
@baumeier do we get a positive review? :)

**Comment by baumeier** [comment]
Only after getting the right results again:

From current master:
```
... ...   ====== Perturbative quasiparticle energies (Hartree) ====== 
... ...    DeltaHLGap = +0.087322 Hartree
... ...   Level =    0 DFT = -14.4223 VXC = -2.0097 S-X = -3.1605 S-C = +0.9148 GWA = -14.6583
... ...   Level =    1 DFT = -14.4176 VXC = -2.0097 S-X = -3.1605 S-C = +0.9134 GWA = -14.6551
```

From earlier slow libint:
```
... ...   ====== Perturbative quasiparticle energies (Hartree) ======
... ...    DeltaHLGap = +0.087322 Hartree
... ...   Level =    0 DFT = -14.4223 VXC = -2.0097 S-X = -3.1605 S-C = +0.9148 GWA = -14.6583
... ...   Level =    1 DFT = -14.4176 VXC = -2.0097 S-X = -3.1605 S-C = +0.9134 GWA = -14.6551
```

From current, fast libint:
```
... ...   ====== Perturbative quasiparticle energies (Hartree) ======
... ...    DeltaHLGap = +0.121476 Hartree
... ...   Level =    0 DFT = -14.4223 VXC = -2.0097 S-X = -3.7453 S-C = +1.3508 GWA = -14.8071
... ...   Level =    1 DFT = -14.4176 VXC = -2.0097 S-X = -4.2012 S-C = +1.7727 GWA = -14.8365
```

File: share/xtp/xml/tolibintorder.xml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0"?>
+<options>
+  <tolibintorder help="Convert .mm matrices from votca to libint order">
+    <basisset>something</basisset>
+    <xyz>something</xyz>
+    <matrix>somehting</matrix>
+    <aux>0</aux>
+    <operator>1</operator>
+  </tolibintorder>
+</options>
```
[the remark this answers is not in the document]
Comment (JensWehner) [reply share/xtp/xml/tolibintorder.xml:10]: ahh we still have to delete that.

**Comment by baumeier** [comment]
Works now!

**Review by baumeier — APPROVED** [review]
Approved from functionality perspective.

**Comment by JensWehner** [comment]
@rubengerritsen final check? You can merge afterwards, thanks for all your work.

**Pull request merged**